drivers/SmartThings/zwave-garage-door-opener/src/ecolink-zw-gdo/init.lua

```diff
@@ -0,0 +1,458 @@
+end
+
+local function endpoint_to_component(device, ep)
+  if ( CONTACTSENSOR_ENDPOINT_NUMBER == ep ) then
+    return CONTACTSENSOR_ENDPOINT_NAME
+  end
+  return GDO_ENDPOINT_NAME
+end
+
+--- Handle BarrierOperator Report
+---
+--- @param driver st.zwave.Driver
+--- @param device st.zwave.Device
+--- @param cmd st.zwave.CommandClass.BarrierOperator.Report
+local function barrieroperator_report_handler(driver, device, cmd)
+  local barrier_event = nil
+  log.info_with
+  (
+    {hub_logs=true}, 
+    "barrieroperator_report_handler> device:%s, cmd:%s", tostring(device), tostring(cmd)
+  )
+  if cmd.args.state == BarrierOperator.state.CLOSED then
+    barrier_event = capabilities.doorControl.door.closed()
+  elseif cmd.args.state == BarrierOperator.state.CLOSING then
+    barrier_event = capabilities.doorControl.door.closing()
+  elseif cmd.args.state == BarrierOperator.state.STOPPED then
+    barrier_event = capabilities.doorControl.door.unknown()
+  elseif cmd.args.state == BarrierOperator.state.OPENING then
+    barrier_event = capabilities.doorControl.door.opening()
+  elseif cmd.args.state == BarrierOperator.state.OPEN then
+    barrier_event = capabilities.doorControl.door.open()
+  end
+
+  if (barrier_event ~= nil) then
+    device:emit_event_for_endpoint(GDO_ENDPOINT_NUMBER, barrier_event)
+  end
+
+end
+
+local function send_open_to_gdo(driver, device, command)
+  if (
+    device:get_latest_state(
+            CONTACTSENSOR_ENDPOINT_NAME,
+            capabilities.tamperAlert.ID,
+            capabilities.tamperAlert.tamper.NAME) == "clear"
+    ) then
+    log.info_with({hub_logs=true}, "<<send_open_to_gdo>> - SUCCESS")
+
+    device:send(BarrierOperator:Set({ target_value = BarrierOperator.state.OPEN }))
+    device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY, function(d)
+      device:send(BarrierOperator:Get({}))
+    end)
+
+  else
+    log.info_with({hub_logs=true}, "<<send_open_to_gdo>> FAILED - SENSOR IN TAMPER")
+
+    device:emit_event_for_endpoint(GDO_ENDPOINT_NUMBER,
+                                  capabilities.doorControl.door.unknown()
+                                  )
+  end
+end
+
+local function send_close_to_gdo(driver, device, command)
+  if (
+    device:get_latest_state(CONTACTSENSOR_ENDPOINT_NAME,
+                            capabilities.tamperAlert.ID,
+                            capabilities.tamperAlert.tamper.NAME) == "clear"
+  ) then
+    log.info_with({hub_logs=true}, "<<send_close_to_gdo>> - SUCCESS")
+
+    device:send(BarrierOperator:Set({ target_value = BarrierOperator.state.CLOSED }))
+    device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY, function(d)
+      device:send(BarrierOperator:Get({}))
+    end)
+else
+  log.info_with({hub_logs=true}, "<<send_close_to_gdo>> FAILED - SENSOR IN TAMPER")
+
+  device:emit_event_for_endpoint(GDO_ENDPOINT_NUMBER,
+                                capabilities.doorControl.door.unknown()
+                                )
+end
+end
```

Comment:
these functions should be consolidated into a factory, like we use with drivers such as zigbee thermostat: https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/main/drivers/SmartThings/zigbee-thermostat/src/init.lua#L277

Reply:
In the example you provided the factory is setup in the main driver, not the subdriver. Would the factory for the GDO open and close commands also be placed in the main driver and used by the subdriver?

Reply:
you would set it up in the sub-driver